/*
 * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
syntax = "proto2";

import "jax_inference_offloading/api/param_mapping.proto";
import "google/protobuf/empty.proto";
import "google/protobuf/any.proto";

message KVPutRequest {
  optional string key = 1;
  optional google.protobuf.Any value = 2;
}
message KVPutResponse {
  optional bool success = 1;
}

message KVGetRequest {
  optional string key = 1;
}
message KVGetResponse {
  optional bool found = 1;
  optional google.protobuf.Any value = 2;
}

message GetNcclIdRequest {}
message GetNcclIdResponse {
  repeated int32 ids = 1;
}

message CreateTransportRequest {
  optional string config_json = 2;
}
message CreateTransportResponse {}

message JaxParallelism {
  optional int32 tp = 1 [default = 1];
  // optional int32 pp = 2 [default = 1];
};
message VllmParallelism {
  optional int32 tp = 1 [default = 1];
  // optional int32 pp = 2 [default = 1];
};
message HandshakeRequest {
  optional string response_topic = 1;
  // optional int32 jax_n_devices = 2;
  reserved 2;
  optional JaxParallelism jax_parallelism = 3;
  optional string model_name = 4;
}
message HandshakeResponse {
  optional TpModelMappingSpecs mapping_specs = 1;
  reserved 2;
  optional JaxParallelism jax_parallelism = 3;
  optional VllmParallelism vllm_parallelism = 4;
}


message StartWeightUpdateRequest {
  optional string mode = 1;
}
message StartWeightUpdateResponse {}

message RolloutConfig {
  optional int32 max_tokens = 1 [default = 64];
  optional float temperature = 2 [default = 0.0];
  optional float top_p = 3 [default = 1.0];
  optional int32 top_k = 4 [default = -1];
  // The number of sequences to generate for the prompt.
  optional int32 num_outputs = 5 [default = 1];
  optional int32 seed = 6 [default = 42];
  repeated int32 stop_token_ids = 7;
}

message TokenIds {
  repeated int32 ids = 1;
}

message Prompt {
  oneof kind {
    string text_prompt = 1;
    string chat_messages_json = 2;
    TokenIds tokenized_prompt = 3;
  }  
}

message InferenceRequest {
  reserved 1, 2, 3;
  repeated Prompt prompts = 6;
  optional string response_topic = 4;
  optional RolloutConfig config = 5;
}

message InferenceResponse {
  message Output {
    optional string generated_text = 1;
    optional TokenIds generated_tokens = 2;
    repeated float generated_token_logps = 3;
    // TODO: Return more logprobs if needed.
    optional TokenIds tokenized_prompt = 4;
  }

  repeated Output outputs = 1;
  reserved 2;
}

message ShutdownRequest {
  optional int32 grace_period = 1 [default = 10];
}
message ShutdownResponse {}

service CouplingController {
  rpc KVPut(KVPutRequest) returns (KVPutResponse) {}

  rpc KVGet(KVGetRequest) returns (KVGetResponse) {}

  rpc GetNcclId(GetNcclIdRequest) returns (GetNcclIdResponse) {}

  rpc CreateTransport(CreateTransportRequest) returns (CreateTransportResponse) {}

  rpc AsyncHandshake(HandshakeRequest) returns (google.protobuf.Empty) {}

  rpc StartWeightUpdate(StartWeightUpdateRequest) returns (StartWeightUpdateResponse) {}

  rpc AsyncInference(InferenceRequest) returns (google.protobuf.Empty) {}

  rpc Shutdown(ShutdownRequest) returns (ShutdownResponse) {}
}
